client/hwlib/src/collectors/cpuid.rs

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use smbioslib;
+
+pub(super) fn get_cpuid(
+    proc_info: &smbioslib::SMBiosProcessorInformation,
+) -> Result<String, Box<dyn std::error::Error>> {
+    let cpu_id = proc_info
+        .processor_id()
+        .map(|id| format!("{:x?}", id))
+        .ok_or("Processor ID not found")?;
+    Ok(cpu_id)
+}
+
+pub(super) fn convert_cpu_codename(cpu_id: &str) -> Result<String, Box<dyn std::error::Error>> {
+    if cpu_id.is_empty() {
+        return Ok(String::from("Unknown"));
+    }
+
+    let cpu_id_bytes: Vec<&str> = cpu_id.split_whitespace().take(3).collect();
```

Comment:
we fetch and treat this value as string, since according to the matching logic, we need to reverse it, the same way it's done in the C3 logic: https://github.com/canonical/hexr/blob/9084e69e88726158f5bd4e1947b3a3916afd33c0/apps/hardware/models.py#L1898-L1904

(The data from C3 is used to match the data that we collect from the client. The hwapi server side just creates the replica of C3 DB, so we need the CPU codename data to be parsed the same way C3 parses it)

Reply:
Ah I think my comment is more low-level than this, it seems strange to call this declaration `cpu_id_bytes` given that we're not handling any `u8`s here. I checked out the the python implementation to see how bytes are used there, but without type annotations, I have no idea what the data being handled is! I think it would be best to go for local consistency here, so as we're splitting into parts, I think it's most natural to call the result `cpu_id_parts`

Reply:
Updated in 186e6d2